File: loader/src/main/java/de/timolia/lactea/loader/module/discovery/DiscoveryClass.java
package de.timolia.lactea.loader.module.discovery;

import de.timolia.lactea.loader.module.LacteaModule;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import javassist.bytecode.annotation.Annotation;
import lombok.Getter;
import lombok.ToString;

/**
 * @author David (_Esel)
 */
@Getter
@ToString(exclude = "byType")
public class DiscoveryClass {
    private final String name;
    private final Annotation[] annotations;
    private final Map<String, Annotation> byType = new HashMap<>();


    public DiscoveryClass(String name, Annotation[] annotations) {
        this.name = name;
        this.annotations = annotations;
        buildTypeIndex();
    }

    private void buildTypeIndex() {
        for (Annotation annotation : annotations) {
            if (byType.put(annotation.getTypeName(), annotation) != null) {
                throw new IllegalStateException("Illegal discovery name=" + name + " annotation=" + annotation);
            }
        }
    }

    public Annotation byType(String search) {
        return byType.get(search);
    }

    public Annotation byType(Class<? extends java.lang.annotation.Annotation> search) {
        return byType(search.getName());
    }

    public Class<? extends LacteaModule> loadClass() throws ClassNotFoundException {
        return (Class<? extends LacteaModule>) Class.forName(name);
    }
}

File: core/src/main/kotlin/de/timolia/lactea/core/discovery/DiscoveryClass.kt
package de.timolia.lactea.core.discovery

import javassist.bytecode.annotation.Annotation as ByteCodeAnnotation

class DiscoveryClass(
    val name: String,
    val annotations: Array<ByteCodeAnnotation>
) {
    private val byType: MutableMap<String, ByteCodeAnnotation> = HashMap()

    init {
        buildTypeIndex()
    }

    private fun buildTypeIndex() {
        for (annotation in annotations) {
            check(byType.put(annotation.getTypeName(), annotation) == null) { "Illegal discovery name=$name annotation=$annotation" }
        }
    }

    fun byType(search: String): ByteCodeAnnotation? {
        return byType[search]
    }

    fun byType(search: Class<out Annotation>): ByteCodeAnnotation? {
        return byType(search.name)
    }

    @Throws(ClassNotFoundException::class)
    fun loadClass(): Class<*> {
        return Class.forName(name)
    }
}
